chore(style-editor):  Created API to define style editor form structure on dotcms/uve

KevinDavilaDotCMS opened the pull request:
https://github.com/user-attachments/assets/67c0f657-49e9-4e5f-9071-2bc830cbea2a



### Usage


```
  // First define the form with the API.
    const blogForm = defineStyleEditorSchema({
        contentType: 'Content',
        sections: [
            {
                title: 'Typography',
                fields: [
                    styleEditorField.checkboxGroup({
                        label: 'New CheckboxGroup',
                        options: [
                            {
                                label: 'Option 1',
                                value: 'option1'
                            },
                            {
                                label: 'Option 2',
                                value: 'option2'
                            }
                        ],
                        defaultValue: {
                            option1: false,
                            option2: false
                        },
                    }),
                    styleEditorField.radio({
                        label: 'Layout',
                        columns: 2,
                        options: [
                            {
                                label: 'Option 1',
                                value: 'option1',
                                imageURL: 'https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcSphn5CRr3MrQUjWWH7ByHWW-lROnVQl4XxYQ&s',
                                width: 119
                            },
                            {
                                label: 'Option 2',
                                value: 'option2',
                                imageURL: 'https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcSphn5CRr3MrQUjWWH7ByHWW-lROnVQl4XxYQ&s'
                            },
                            {
                                label: 'New Option',
                                value: ''
                            },
                            {
                                label: 'New Option',
                                value: 'new-option'
                            }
                        ],
                    })
                ]
            }
        ]
    })
    
    // Then use the helper hook.

    useStyleEditorSchemas([blogForm])
    // This hook internally use a exposed function called registerStyleEditorSchemas()
    // In Angular, we can use this function directly. We can create a service for this too, but maybe is a lot of code only for that, but we can recheck this later 
```

This PR fixes: #33809

kevindaviladev committed:
WIP - Need to change to other branch to hotfix

kevindaviladev committed:
Enhance style editor functionality by adding type exports and improving input handling

- Added export for  in  to define valid input types.
- Updated  function in  to ensure type safety for  and .
- Introduced comprehensive unit tests in  to validate the behavior of various field types, including input, dropdown, and radio fields.
- Enhanced documentation in  to clarify usage and experimental status of style editor functions.

kevindaviladev committed:
Changed version of react and react-dom to be sync with main

kevindaviladev committed:
Refactor style editor types for improved clarity and type safety

- Renamed  to  and  to  for better context.
- Updated type references in  and  to use the new type names, enhancing type safety in the style editor functionality.

kevindaviladev committed:
Add useStyleEditorSchemas hook and related tests for style editor functionality

- Introduced the  hook to register style editor forms with the UVE editor.
- Added comprehensive unit tests for the hook to validate its behavior with various form schemas.
- Updated the public API to include a new action for registering style schemas in the editor.
- Enhanced the style editor's public interface to support the new functionality.

kevindaviladev committed:
Refactor style editor normalization and enhance type safety

- Updated the normalization process for field and section definitions to ensure all type-specific properties are encapsulated within a  object, promoting consistency in the schema format sent to UVE.
- Enhanced type safety for input fields by introducing a new  type, ensuring that  types align with .
- Improved documentation to clarify the normalization process and the structure of the resulting schemas.
- Adjusted the public API to reflect these changes, ensuring a seamless integration for developers using the style editor.

kevindaviladev committed:
Enhance style editor functionality with radio field support and layout options

- Added handling for radio-specific properties in the normalization process, allowing for dynamic column configurations.
- Updated the structure of normalized fields to simplify the schema format.
- Improved test cases for style editor forms to reflect changes in field definitions and layout options.
- Enhanced documentation to clarify the usage of columns in radio fields and their impact on layout.

kevindaviladev committed:
Refactor style editor API to rename form definition function

- Updated the  function to  for improved clarity and consistency in naming.
- Adjusted all references in tests and documentation to reflect the new function name.
- Ensured that the public API remains consistent with the updated naming convention, enhancing developer experience.

kevindaviladev committed:
Upgrade nextjs and react version to avoid vulnerability check on deploy pipeline. Same version as main

zJaaal commented on core-web/libs/sdk/uve/src/lib/style-editor/types.ts:
```diff
@@ -0,0 +1,616 @@
+          placeholder?: string;
+          /** Optional default value (must be a string when inputType is 'text') */
+          defaultValue?: string;
+      });
```
I think the first `|` is not needed and also we should make use of the types we have above instead of pure strings

zJaaal commented on core-web/libs/sdk/uve/src/lib/style-editor/types.ts:
```diff
@@ -0,0 +1,616 @@
+              /** Optional height of the image in pixels */
+              height?: number;
+          }
+    >;
```
I think the first `|` is not needed

zJaaal commented on core-web/libs/sdk/uve/src/lib/style-editor/types.ts:
```diff
@@ -0,0 +1,616 @@
+    /** Discriminator: must be 'checkboxGroup' */
+    type: 'checkboxGroup';
+    /** Array of selectable options. Can be strings or objects with label and value properties */
+    options: Array<string | { label: string; value: string }>;
```
same here for the label/value

KevinDavilaDotCMS replied:
Added!

kevindaviladev committed:
Enhance style editor type definitions for improved option handling

- Introduced new interfaces for option objects, including  and , to standardize option structures across dropdowns, radio buttons, and checkboxes.
- Updated existing field interfaces to utilize these new types, enhancing type safety and clarity in option handling.
- Added a new  interface to encapsulate field-specific properties, streamlining the configuration process for various field types.
- Improved documentation to clarify the usage and structure of the new option types and field schema configuration.

KevinDavilaDotCMS commented on core-web/libs/sdk/uve/src/lib/style-editor/types.ts:
```diff
@@ -0,0 +1,616 @@
+        /** Optional maximum value constraint for number inputs */
+        max?: number;
+        /** Optional regex pattern for text input validation */
+        pattern?: string;
```
Added!

kevindaviladev committed:
Enhance style editor type definitions with checkbox default value support

- Introduced a new type, StyleEditorCheckboxDefaultValue, to standardize the default value structure for checkbox groups.
- Updated existing interfaces to utilize the new type, improving type safety and clarity in checkbox handling.
- Enhanced documentation to provide examples and clarify the usage of the new checkbox default value type.

kevindaviladev committed:
Enhance useStyleEditorSchemas tests for infinite loop prevention

- Added tests to ensure the useStyleEditorSchemas hook does not cause infinite loops when inline objects are passed on each render.
- Demonstrated the importance of stable references to prevent unnecessary re-registrations during re-renders.
- Updated existing test cases to validate the expected behavior of the hook under various rendering scenarios.

kevindaviladev committed:
Merge branch 'main' into 33809-task-createapi-to-define-style-editor-form-structure-on-dotcmsuve

kevindaviladev committed:
Refactor style schema handling in Edit EMA Editor

- Updated the style schema interface to use StyleEditorFormSchema for improved type safety.
- Modified the handling of style schemas in the EditEmaEditorComponent to align with the new structure.
- Enhanced the display of style schemas in the dot-uve-palette component for better usability.
- Adjusted related tests to reflect changes in schema structure and ensure correct functionality.

kevindaviladev committed:
Fixed lint issue on edit-ema-editor.component.ts

rjvelazco commented on core-web/libs/sdk/react/src/lib/next/hooks/useStyleEditorSchemas.ts:
```diff
@@ -0,0 +1,14 @@
+ * @param forms - Array of style editor form schemas to register
+ * @returns void
+ */
+export const useStyleEditorSchemas = (styleEditorForms: StyleEditorFormSchema[]): void => {
```
`@experimental` until official release

KevinDavilaDotCMS commented on core-web/libs/sdk/uve/src/lib/style-editor/internal.ts:
For the internal things i think is ok to dont use prefix
